### ISSUE TITLE
Adding a GAE example

### ISSUE THREAD

**Issue by thegodone** [akensert/molgraph#7]
I wonder if you can implement the basic softmax "innerProduct" of the embedding vector representation of the Graph, this is the baseline of decoder in initial GAE code.

https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.InnerProductDecoder.html#torch_geometric.nn.models.InnerProductDecoder

so you can use any encoder like GIN or GAT or GT to get Z embedding and retrieve back Adj using : https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.GAE.html#torch_geometric.nn.models.GAE

My question is also how to do so with the graphTensor "data type" ?

**Comment by thegodone** [comment]
This is very cool, I did not look at all you codes indeep. thanks.

Le mer. 7 juin 2023 à 10:37, Alexander Kensert ***@***.***> a
écrit :

**Comment by akensert** [comment]
What is the purpose of GAE? Simply pretraining?

I'm implementing the sparse version ("forward" of the GAE you linked), and not the dense ("forward_all" of the GAE you linked). Perhaps the "forward_all" should also be implemented. My understanding:

- The sparse version simply performs a dot product between incident nodes, and the prediction should be `1`. And as 'negative' examples we can randomly shuffle 'edge_dst' and those should be predicted as `0`. We are using binary crossentropy.
- The dense version performs a matrix multiplication of `z` (of shape (`n_nodes x 1`)). The result is a adjacency matrix (with values within the range [0, 1]). My intuition here is that we have alot of negative examples as usually the adjacency matrix is sparse (mostly nodes/atoms are not connected to other nodes/atoms). To make the sparse version more similar we could sample the graph with negative examples (edges) times `k`.

I'm not really sure if we need to implement both. In training (so for the purpoise of pretraining an encoder), they do very similar things. However, I miss out on some application. Let me know what you think.

Also, there exist some interesting losses in regard to link prediction (seems to not exist in the documentation): [see here](https://github.com/akensert/molgraph/blob/main/molgraph/losses/link_losses.py)

`y_true` should be positive examples and `y_pred` negative examples (if I remember correctly).

In the GAE example you linked above, they seem to use the `LinkBinaryCrossentropy`. Though there are other options too (never really tried this out).

`

**Comment by thegodone** [comment]
correct First GAE is for pretrained at least good embedding. Sparse can be an option yes. But Dense is more challenging task with an output sigmoid (Dense) that may help to have a more robust AE than Sparse simpler task. I think yes LinkBinaryCrossentropy makes sense. I think also masking atom/bond can be very useful also to try to reconstructed based on partial corrupted input Graph.

**Comment by thegodone** [comment]
One independent question: 
this is sequence "EdgeConv(), NodeReadout(), GINConv()" is equivalent to DGIN (1 conv) and this second sequence
"EdgeConv(), NodeReadout(), MPNNConv()" is for DMPNN (1 conv).

Is it correct ?

for almost all the method you have the Readout() to merge into a dense layer. is it the same for GTConv ?

**Comment by akensert** [comment]
_"this is sequence "EdgeConv(), NodeReadout(), GINConv()" is equivalent to DGIN (1 conv) and this second sequence "EdgeConv(), NodeReadout(), MPNNConv()" is for DMPNN (1 conv)._

_Is it correct ?"_

- For DGIN (using layers): 
> - `EdgeConv()` for `t` number of steps; 
> 2. `NodeReadout()`; 
> 3. `GINConv()` for `k` number of steps;
- For DMPNN (using layers): 
> 1. `EdgeConv()` for `t` number of steps; 
> 2. `NodeReadout()`; 
> 3. A final linear transformaton e.g. `FeatureProjection(feature='node_feature', ...)`.

So for DGIN, you do a number of "edge convs" and then a number of "node convs"; for DMPNN you don't do the "node convs". Of course you can do that if you'd like :), but it is not implemented in the  `DMPNN` model. I based this on litterature. I forgot exact paper but you can find it in the documentation. 

_for almost all the method you have the Readout() to merge into a dense layer. is it the same for GTConv ?_

**Yes.**

**Comment by akensert** [comment]
Btw, I just fixed a bug. Apparently I passed a keras.initializer instance to __init__ which made all Conv layers (assuming `kernel_initializer` is not explicitly passed) to have the same initial values. I've now fixed that. Feel free to update to version `0.3.5`. Sorry for the inconvenience.

**Comment by akensert** [comment]
@thegodone Regarding EdgeConv (and MPNNConv): I will modify the layers slightly. Specifically, I will remove the `tie_layer` argument and replace it with a `gru_cell` argument. I might also make `update_mode` default to `gru` instead of `dense`. So if it is desired to share the gru update between layers, then a GRUCell can be defined outside and passed to all the layers. So avoid using `tie_layer` as I will remove it in the near future.

As long as your code look something like this (for DGIN):
```python
gnn_model = tf.keras.Sequential([
       tf.keras.Input(type_spec=graph_tensor.unspecific_spec),
       molgraph.layers.EdgeConv(16, update_mode='gru'),   # or update_mode='dense'
       molgraph.layers.EdgeConv(16, update_mode='gru'),   # or update_mode='dense'
       # ... optionally more EdgeConv layers here
       molgraph.layers.NodeReadout(),  
       molgraph.layers.GINConv(32),
       molgraph.layers.GINConv(32),
       # ... optionally more GINConv layers here
       molgraph.layers.Readout(),
       # ... optionally some dense layers here
])
```
Or for DMPNN:
```python
gnn_model = tf.keras.Sequential([
       tf.keras.Input(type_spec=graph_tensor.unspecific_spec),
       molgraph.layers.EdgeConv(16, update_mode='gru'),   # or update_mode='dense'
       molgraph.layers.EdgeConv(16, update_mode='gru'),   # or update_mode='dense'
       # ... optionally more EdgeConv layers here
       molgraph.layers.NodeReadout(),  
       molgraph.layers.FeatureProjection(feature='node_feature'),
       molgraph.layers.Readout(),
       # ... optionally some dense layers here
])
```

It should be fine (code should not break)

**Comment by thegodone** [comment]
I move to 0.3.6 thanks for the fixes.

**Comment by thegodone** [comment]
To be 100% sure: EdgeConv == DMPNN == ChemProp correct ?

**Comment by akensert** [comment]
> To be 100% sure: EdgeConv == DMPNN == ChemProp correct ?

Yes. `EdgeConv` is used in both DGIN and DMPNN based on [this paper](https://www.mdpi.com/1420-3049/26/20/6185). You are likely the first one to really try it out (assuming you are going to try it out), so feel free to share how it's doing.

**Comment by thegodone** [comment]
I am using it I still don't see the performances comparable to the paper. but I will let you know for sure.

So far I tried 11 of your Graphs to make a proper benchmark on ESOL with your random split.

I will let you know soon my findings.

**Comment by akensert** [comment]
> I am using it I still don't see the performances comparable to the paper. but I will let you know for sure.

Hmm okay. Could be some subtle things like different initializers. Or different train/valid/test splits. How difference is the performance? Hopefully not, but could also be some bug in the implementation.

**Comment by thegodone** [comment]
so comparing results on test ESOL dataset

type | rmse
-- | --
AttentiveFP | 0.609508
GAT | 0.623783
GIN | 0.597040
GINE | 0.607946
GT | 0.603691
GCN | 0.580124
GatedGCN | 0.644287
MPNN | 0.670782
GMM | 0.589558
DGIN | 3.482894
DMPNN | 4.480392

GINE is really a good improvement compare to GIN I guess DGINE could be good. Anyway only one convolution of DGIN and DMPNN looks wrong is my code correct ?

class DGIN(tf.keras.Model):
    def __init__(self,embdim, **kwargs):
        super(DGIN, self).__init__(**kwargs)
        
        self.EdgeConv = EdgeConv(embdim)
        self.NodeReadout = NodeReadout(apply_transform=True)
        self.GIN = GINConv()
        self.Readout = Readout()
        
    def call(self, inputs):
        x = self.EdgeConv(inputs)
        x = self.NodeReadout(x)
        x = self.GIN(x)
        return self.Readout(x)
    
    
    
class DMPNN(tf.keras.Model):
    def __init__(self,embdim, **kwargs):
        super(DMPNN, self).__init__(**kwargs)
        
        self.EdgeConv = EdgeConv(embdim)
        self.NodeReadout = NodeReadout(apply_transform=True)
        self.FP = FeatureProjection(embdim,feature='node_feature')
        self.Readout = Readout()
        
    def call(self, inputs):
        x = self.EdgeConv(inputs)
        x = self.NodeReadout(x)
        x = self.FP(x)
        return self.Readout(x)

**Comment by akensert** [comment]
@thegodone I'm testing out the EdgeConv in a model. I get this error:
```python
No unary variant unary_op function found for op ZEROS_LIKE Variant type_name: RaggedTensorVariant for device type: GPU
```
I need to look into this. But you managed to at least run your model?

**Comment by thegodone** [comment]
yes on CPU. 

	type	rmse
0	AttentiveFP	0.583507
1	GAT	0.622939
2	GIN	0.614117
3	GINE	0.587490
4	GT	0.574342
5	GCN	0.568994
6	GatedGCN	0.638000
7	MPNN	0.600579
8	GMM	0.635617
9	DGIN	3.134166
10	DMPNN	5.255269

**Comment by thegodone** [comment]
ESOL is not the dataset they used it much larger and more noisy

**Comment by akensert** [comment]
> ESOL is not the dataset they used it much larger and more noisy

Sorry I meant the molgraph manuscript. I get a RMSE of around 0.64 with GAT, but in the tutorial I get 0.54.

**Comment by thegodone** [comment]
this is only one split ;-)

**Comment by akensert** [comment]
> @thegodone I'm testing out the EdgeConv in a model. I get this error:
> 
> ```python
> No unary variant unary_op function found for op ZEROS_LIKE Variant type_name: RaggedTensorVariant for device type: GPU
> ```
> 
> I need to look into this. But you managed to at least run your model?
> 
> EDIT: It seems to run on a CPU.
> 
> EDIT: Seems like `tf.map_fn` does not currently work well on GPU with `tf.RaggedTensor`s. I have to figure out a way to implement it differenly.

I've now fixed this. Should now run fine on GPU as well :) Implemented it with `tf.TensorArray` instead of `tf.map_fn`.   (will soon push the updates)

**Comment by akensert** [comment]
Regarding GAE:

I've implemented a GAE that seems to work fine, however, I have some issues with the GVAE. For some reason there is  no convergence on the reconstruction loss when using kl loss. Removing the kl loss, it converges fine. Although the pytorch implementation doesn't do that, I wonder if I need to add a coefficient to reduce the influence of the kl loss?

**Comment by thegodone** [comment]
We can try with a coefficient in the loss (strange). The question is converge how well ?

**Comment by thegodone** [comment]
I cannot make running your model example on cpu. kernel is kill each time
```

model = tf.keras.Sequential([
    keras.layers.Input(type_spec=type_spec),
    node_preprocessing,
    edge_preprocessing,
    EdgeConv(units=128),
    EdgeConv(units=128),
    EdgeConv(units=128),
    EdgeConv(units=128),
    NodeReadout(apply_transform=True),
    GINConv(use_edge_features=True),
    GINConv(use_edge_features=True),
    GINConv(use_edge_features=True),
    Readout(),
    keras.layers.Dense(1024, 'relu'),
    keras.layers.Dense(1024, 'relu'),
    keras.layers.Dense(y_train.shape[-1])
])
```

**Comment by thegodone** [comment]
yes working with 3.8

**Comment by akensert** [comment]
> yes working with 3.8, but it is very very slow on the cpu... compare to other models

What is very very slow? Like 2 times slower? 10 times slower?

I'm expecting it to be somewhat slower, but not like 10 times slower. As there is a lot of computation. many steps of convolutions. 

Btw, do work on your fork? Or do you install via pip? 0.3.8 is not pip installable yet.

**Comment by thegodone** [comment]
I do git pull / python setup.py install / I change the setup file to use
tensorfow-macos

Le jeu. 8 juin 2023 à 18:05, Alexander Kensert ***@***.***> a
écrit :

**Comment by thegodone** [comment]
I did a mistake it was on mps so why it was so slow on cpu it works and it fast!

**Comment by thegodone** [comment]
Do you have a linkedin account ?

**Comment by thegodone** [comment]
Sorry one more question: in the code you put 4 Edge and 3 GIN do we need always an extra Edge vs number of GIN ?

**Comment by akensert** [comment]
> Sorry one more question: in the code you put 4 Edge and 3 GIN do we need always an extra Edge vs number of GIN ?

I don't think so :) I just randomly picked 4 EdgeConv and 3 GINConv

**Comment by akensert** [comment]
> Indeed, I play on models and when I plug two models types directed + undirected as two type of convolutions in parallel : A: input -> DGIN(input) -> DNN(64,32,1) B: input -> GIN(input) -> DNN(64,32,1) and the result is [y,y] In this model DGIN is diverse a lot rmse >3.4 and GIN rmse is 0.6 But DGIN alone is working well so I was thinking that something is not compatible between Directed & Undirected

I don't quite follow. How are you running it in parallel, can you show me some code?

**Comment by thegodone** [comment]
```
class DGIN(tf.keras.layers.Layer):
    def __init__(self,embdim, steps, **kwargs):
        super(DGIN, self).__init__(**kwargs)
        self.steps = steps
        self.EdgeConv = [EdgeConv(embdim) for _ in range(self.steps)]
        self.NodeReadout = NodeReadout(apply_transform=True)
        self.GIN = [GINConv(embdim, use_edge_features=True) for _ in range(self.steps)]
        
    def call(self, inputs):
        x = inputs
        for i in range(self.steps):
            x = self.EdgeConv[i](x)
     
        x = self.NodeReadout(x)
        
        for i in range(self.steps):
            x = self.GIN[i](x)
        
        return x

class Mymodel(tf.keras.Model):
    def __init__(self, embdim, **kwargs):
        super(Mymodel, self).__init__(**kwargs)
        # undirected graphs special execption for AttFP for the readout function!
        #self.node_preprocess = node_preprocessing
        #self.edge_preprocess = edge_preprocessing
        self.graphsconv = [GINConv(embdim, use_edge_features=True), 
                           DGIN(embdim)]
        self.par = len(self.graphsconv)
        self.MolReadout = [Readout() for _ in range(self.par)]
                    
        self.layer_normalization = [tf.keras.layers.LayerNormalization() for _ in range(self.par)]
        self.dense1 = [tf.keras.layers.Dense(embdim) for _ in range(self.par)]
        self.leaky_relu1 = tf.keras.layers.LeakyReLU()
        self.dense2 = [tf.keras.layers.Dense(32) for _ in range(self.par)]
        self.leaky_relu2 = tf.keras.layers.LeakyReLU()
        self.dense3 = [tf.keras.layers.Dense(1) for _ in range(self.par)]
        self.concate = tf.keras.layers.Concatenate()
        
    def call(self, inputs):       
        r = []
        for i in range(self.par):
            xi = self.graphsconv[i](inputs)
            xi = self.MolReadout[i](xi)  
            xi = self.layer_normalization[i](xi)                                 
            xi = self.dense1[i](xi)
            xi = self.leaky_relu1(xi)
            xi = self.dense2[i](xi)
            xi = self.leaky_relu2(xi)
            r.append(self.dense3[i](xi))
                                 
        return self.concate(r)
```

**Comment by thegodone** [comment]
@akensert did you already made this changes ?  "Regarding EdgeConv (and MPNNConv): I will modify the layers slightly. Specifically, I will remove the tie_layer argument and replace it with a gru_cell argument. I might also make update_mode default to gru instead of dense. So if it is desired to share the gru update between layers, then a GRUCell can be defined outside and passed to all the layers. So avoid using tie_layer as I will remove it in the near future." ?

**Comment by thegodone** [comment]
my last results

  | type | rmse
-- | -- | --
GINE | 0.593514
DGIN | 0.568289
GIN | 0.613843
GMM | 0.600482
GT | 0.786425
GCN | 0.632253
MPNN | 0.660916
DMPNN | 0.608485
GAT | 0.624454

**Comment by akensert** [comment]
I can try to push what I have to the develop branch tomorrow. 

It seems like the coefficient is necessary to make sure the reconstruction loss decreases. I learned that it is important to properly "balance" the reconstruction loss vs. kl loss (using a coefficient `beta` for the kl loss). It has even been recommended to set `beta` to 0 at the beginning of training, and then increase it in small steps.

About the dataset: I forgot which datasets, but it was something like a thousand molecules (which is not much).

**Comment by thegodone** [comment]
Those atoms are equivalent for a chemist that is nice that is it the same for the computer. There should be a way to assert the permutation of those two atoms does not affect the reconstruction result. Maybe Laplacian can be used here to get idea of the primary sequence ? This is just an idea.Envoyé de mon iPhoneLe 14 juin 2023 à 09:57, Alexander Kensert ***@***.***> a écrit :﻿
I found an interesting problem, considering this molecule:

I implemented a negative graph sampler that makes sure that when shuffling edges_{i} that edges_{new_i} != edges_{i}. However, there are edges that are identical still (note: same problem exist for dense adjacency representation). For example 0->1 and 2->1. With identical I mean there is no way of distinguishing them by looking at the (aggregated) node features (and likewise the dot product of the aggregated node features). Any solution to this? Or just leave it? The issue is that we want our model to predict 0 and 1 at the same time, given that the edges are shuffled in such a way (this occurs frequently actually) that a new negative edge is (aggregated) node feature-wise a positive edge. Does it make sense?

**Comment by akensert** [comment]
@thegodone autoencoders are now added (but sort of work in progress)

**Comment by akensert** [comment]
> Those atoms are equivalent for a chemist that is nice that is it the same for the computer. There should be a way to assert the permutation of those two atoms does not affect the reconstruction result. Maybe Laplacian can be used here to get idea of the primary sequence ? This is just an idea. what do you think ?

Interesting idea. Something like that would be great. Not sure how others deal with this, if they do at all..

**Comment by thegodone** [comment]
the basic GAE code is quite old for AI timeline. Pretty sure there is something smarter today to improve this point. I don't think they care / dig to much at the beginning of GAE... nice that you find it and consider it as a bug.

**Comment by akensert** [comment]
> UPDATE: Will add docstring with example after lunch.

Examples/documentation now added, see [here](https://molgraph.readthedocs.io/en/latest/api/models.html#pretraining).

It would be great to further improve the GAE/GVAE, before closing this one!

**Comment by thegodone** [comment]
should we look at this method for improving the autoencoder ?
we should think like puzzle to avoid the chemical equivalent atoms ambiguity:
https://github.com/hhaji/funqg https://pubs.acs.org/doi/10.1021/acs.jcim.3c00445 there is also a arxiv link https://arxiv.org/abs/2207.08597 there are also works on 3D fragments in MIT papers 2021.
So the main idea is to create metanodes to simplify the task. what do you think ?
